android/runtime/v8/src/native/TypeConverter.cpp

```diff
@@ -540,6 +540,29 @@ Local<Array> TypeConverter::javaArrayToJsArray(Isolate* isolate, JNIEnv *env, ji
 	return jsArray;
 }
 
+Local<Array> TypeConverter::javaArrayToJsArray(Isolate* isolate, jbyteArray javaByteArray)
+{
+	JNIEnv *env = JNIScope::getEnv();
+	if (env == NULL) {
+		return Local<Array>();
+	}
+	return TypeConverter::javaArrayToJsArray(isolate, env, javaByteArray);
+}
+
+Local<Array> TypeConverter::javaArrayToJsArray(Isolate* isolate, JNIEnv *env, jbyteArray javaByteArray)
+{
+	int arrayLength = env->GetArrayLength(javaByteArray);
+	Local<Array> jsArray = Array::New(isolate, arrayLength);
+
+	Local<Context> context = isolate->GetCurrentContext();
+	jbyte *arrayElements = env->GetByteArrayElements(javaByteArray, 0);
+	for (int i = 0; i < arrayLength; i++) {
+		jsArray->Set(context, (uint32_t) i, Integer::New(isolate, arrayElements[i]));
+	}
+
+	return jsArray;
```

Comment:
I looked at our "v8.h" header file. We should be able to use `memcpy()` to copy the bytes over which should be significantly faster. Especially for large arrays. Would you mind giving this a go please?
```cpp
int byteCount = env->GetArrayLength(javaByteArray);
Local<Array> jsArray = Array::New(isolate, byteCount);
if (byteCount > 0) {
	void* sourcePointer = (void*)env->GetByteArrayElements(javaByteArray, 0);
	void* destinationPointer = jsArray->GetContents().Data();
	memcpy(destinationPointer, sourcePointer, byteCount);
}
return jsArray;
```

An even better way of doing this is to "pin" the Java byte array into memory and wrap it with the V8 ArrayBuffer via the `Array::New(Isolate*, void*, size_t)` API... **but**... **unfortunately**... Android JNI does not guarantee it will pin it in memory and may make a copy instead. So, this idea is out. _(Bummer.)_
https://developer.android.com/training/articles/perf-jni#primitive-arrays

Reply:
I don't think this will work, because the `GetContents().Data()` is an `ArrayBuffer` API, not an `Array` API. So I could conceivably return an `ArrayBuffer` here instead of an `Array`. I alluded to it in the PR body, but basically the most efficient mechanism here is to make use of ByteBuffer in java and shared the underlying memory with the Uint8Array/TypedArray in the JS engine. But that'd require some changes to our Java APIs to use the `ByteBuffer` in place of passing a `byte[]` around.

I'll see if I can't just swap to have it return an ArrayBuffer under the hood. It'll require tweaks in other places, but that may be a better fit.

Reply:
Ok, so looks like the Web `Blob` actually has an API for this! https://developer.mozilla.org/en-US/docs/Web/API/Blob/arrayBuffer

They use method called `arrayBuffer()` that returns a `Promise` that resolves to an `ArrayBuffer`. Given that, it probably makes sense to try and align here. The obvious issue right now is our lack of notice Promise APIs (see #10554)

So in the interim, perhaps a `Ti.Blob.toArrayBuffer()` which directly/synchronously returns an `ArrayBuffer` (rather than the initial PR proposal of `toUint8Array()` returning an `Uint8Array` sync); and then add a JS level shim to add the `Ti.Blob.arrayBuffer()` that returns a `Promise` and uses the sync `toArrayBuffer()` under the hood.

Reply:
We could still try to make use of pinned array elements?
```C++
size_t byteCount = env->GetArrayLength(javaByteArray);
if (byteCount == 0) {
  return ArrayBuffer::New(isolate, 0);
}

jboolean isCopy = JNI_FALSE;
void* src = (void*)(env->GetByteArrayElements(javaByteArray, &isCopy));

if (isCopy == JNI_FALSE) {
  return ArrayBuffer::New(isolate, src, byteCount);
}

Local<Array> jsArray = ArrayBuffer::New(isolate, byteCount);
void* dst = jsArray->GetContents().Data();
memcpy(dst, src, byteCount);
return jsArray;
```

Reply:
First, the code I posted is missing a call to `ReleaseByteArrayElements()`. This is because`GetByteArrayElements()` increments the reference count. So, we must call the release method to decrement it and avoid a memory leak.

If we were to take advantage of "pinning", then we shouldn't release the Java byte array reference until the V8 array object has been garbage collected.

Reply:
It's possible to use `GetPrimitiveArrayCritical` to guarantee pinned access. However, using this locks the GC while the array is being used.

I'm not sure if ART implements a different behaviour, where GC is still possible but "moving" GCs (compacting) are not.
https://android.googlesource.com/platform/art/+/refs/heads/master/runtime/jni/jni_internal.cc#2099

Reply:
I think the correct long-term move here it to use `ByteBuffer` and https://docs.oracle.com/javase/7/docs/technotes/guides/jni/spec/functions.html#NewDirectByteBuffer

I opened https://jira.appcelerator.org/browse/TIMOB-28039 to track this.

Basically the underlying `ArrayBuffer` in the engine can give us a pointer to the memory and we can allocate a direct `ByteBuffer` to expose that to Java. On iOS, we can create an `ArrayBuffer` and have an `NSData` point at the same memory pool.

So for this PR, I'm really trying not to break our APIs but just do a copy of the `byte[]` to generate a JS `ArrayBuffer` as a quicker means of "dumping" bytes rather than traversing the bridge back and forth to read/write byte-by-byte. The current `Ti.Buffer` is written around a `byte[]` and `getBytes()` is a "module" api.

But I *definitely* think we can use memory sharing long-term to underpin `Ti.Buffer` and avoid traversing the bridge at all for reads/writes and still have both the native and JS sides referring to the same bytes.